pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+			// There's a possible error here:
+			// - NC sees a new Node and assigns CIDRs X,Y.. to it,
+			// - Update Node call fails with a timeout,
+			// - Node is updated by some other component, NC sees an update and
+			//   assigns CIDRs A,B.. to the Node,
+			// - Both CIDR X,Y.. and CIDR A,B.. are marked as used in the local cache,
+			//   even though Node sees only CIDR A,B..
+			// The problem here is that in in-memory cache we see CIDR X,Y.. as marked,
+			// which prevents it from being assigned to any new node. The cluster
+			// state is correct.
+			// Restart of NC fixes the issue.
+			if len(newNode.Spec.PodCIDRs) == 0 {
+				return ra.AllocateOrOccupyCIDR(newNode)
+			}
+			return nil
+		}),
+		DeleteFunc: controllerutil.CreateDeleteNodeHandler(ra.ReleaseCIDR),
+	})
+
+	return ra, nil
+}
+
+func (r *multiCIDRRangeAllocator) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting Multi CIDR Range allocator")
+	defer klog.Infof("Shutting down Multi CIDR Range allocator")
+
+	if !cache.WaitForNamedCacheSync("multi_cidr_range_allocator", stopCh, r.nodesSynced, r.clusterCIDRSynced) {
+		return
+	}
+
+	// raWaitGroup is used to wait for the RangeAllocator to finish the goroutines.
+	var raWaitGroup sync.WaitGroup
+
+	for i := 0; i < cidrUpdateWorkers; i++ {
+		raWaitGroup.Add(1)
+		go func() {
+			defer raWaitGroup.Done()
+			r.worker(stopCh)
+		}()
+	}
+
+	raWaitGroup.Wait()
+
+	<-stopCh
+}
+
+func (r *multiCIDRRangeAllocator) worker(stopChan <-chan struct{}) {
+	for {
+		select {
+		case workItem, ok := <-r.nodeCIDRUpdateChannel:
+			r.lock.Lock()
+			if !ok {
+				klog.Error("Channel nodeCIDRUpdateChannel was unexpectedly closed")
+				return
+			}
+			if err := r.updateCIDRsAllocation(workItem); err != nil {
+				// Requeue the failed node for update again.
+				r.nodeCIDRUpdateChannel <- workItem
+			}
+			r.lock.Unlock()
+		case <-stopChan:
+			klog.Infof("MultiCIDRRangeAllocator worker is stopping.")
+			return
+		}
+	}
+}
+
+// createClusterCIDRHandler creates clusterCIDR handler.
+func createClusterCIDRHandler(f func(ccc *networkingv1alpha1.ClusterCIDR) error) func(obj interface{}) {
+	return func(originalObj interface{}) {
+		ccc := originalObj.(*networkingv1alpha1.ClusterCIDR)
+		if err := f(ccc); err != nil {
+			utilruntime.HandleError(fmt.Errorf("error while processing ClusterCIDR Add/Delete: %w", err))
+		}
+	}
+}
+
+// needToAddFinalizer checks if a finalizer should be added to the object.
+func needToAddFinalizer(obj metav1.Object, finalizer string) bool {
+	return obj.GetDeletionTimestamp() == nil && !slice.ContainsString(obj.GetFinalizers(),
+		finalizer, nil)
+}
+
+func (r *multiCIDRRangeAllocator) syncClusterCIDR(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing clusterCIDR request %q (%v)", key, time.Since(startTime))
+	}()
+
+	clusterCIDR, err := r.clusterCIDRLister.Get(key)
+	if apierrors.IsNotFound(err) {
```

Comment:
can there be some scenario we need to handle if we get here? some kind of force delete we don't see the update setting the deletion timestamp?